### PR TITLE
Search event

### PR DESCRIPTION
**Pull request by minhchau273**
- [x] [f] Search upcoming events
- [x] [t] Test searching events

@seri Check it out 😄

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -1,6 +1,11 @@
 class EventsController < ApplicationController
   def index
     @events = Event.upcoming
+    if (@keyword = params["search"]).present?
```
Comment (seri) [inline app/controllers/events_controller.rb:4]: Implement this logic via a class method `Event.search(keyword)`

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
+      keywords = @keyword.downcase.split.join("|")
+      where_clause = "lower(name) similar to '%(#{keywords})%'"
```
Comment (seri) [inline app/controllers/events_controller.rb]: Capitalize SQL keywords.

File: app/views/events/index.html.slim [hunk 3]
```diff
@@ -3,9 +3,13 @@
     h2
       |  Discover upcoming events 
     form
-      input.form-control type="text" name="search" placeholder="Search for events, shows, courses..."
+      input.form-control type="text" name="search" placeholder="Search for events, shows, courses..." value="#{@keyword}" onfocus="this.value = this.value;" autofocus="true"
```
Comment (seri) [inline app/views/events/index.html.slim:6]: Interesting. Please remind me to revisit this in case I forget.

File: app/views/events/index.html.slim [hunk 4]
```diff
@@ -3,9 +3,13 @@
...
 .container-fluid.events-list
-  - @events.in_groups_of(2) do |group|
-    .row
-      - group.compact.each do |event|
-        = render partial: "card", locals: { event: event }
+  - if @events.count == 0
```
Comment (seri) [inline app/views/events/index.html.slim]: `if @events.empty?`

File: features/step_definitions/view_upcoming_events_steps.rb [hunk 5]
```diff
@@ -21,3 +24,32 @@
...
+When "I input a dummy keyword" do
+  fill_in "search", with: DUMMY_KEYWORD
+end
+
+Then "I press enter to search" do
+  find_field("search").native.send_key(:enter)
+end
+
+And "I can see that dummy keyword in the search box" do
+  find_field("search").value.should eq DUMMY_KEYWORD
```
Comment (seri) [inline features/step_definitions/view_upcoming_events_steps.rb]: Use new `expect().to` syntax.

File: features/step_definitions/view_upcoming_events_steps.rb [hunk 6]
```diff
@@ -21,3 +24,32 @@
...
+
+When "I input a dummy keyword" do
+  fill_in "search", with: DUMMY_KEYWORD
+end
+
+Then "I press enter to search" do
+  find_field("search").native.send_key(:enter)
+end
+
+And "I can see that dummy keyword in the search box" do
```
Comment (seri) [inline features/step_definitions/view_upcoming_events_steps.rb]: Consider use `"I can see that keyword in the search box"` the same for both cases no results and same results.

File: features/step_definitions/view_upcoming_events_steps.rb [hunk 7]
```diff
@@ -1,10 +1,13 @@
 Capybara.default_selector = :xpath
 
+KEYWORD = "1"
```
Comment (seri) [inline features/step_definitions/view_upcoming_events_steps.rb:3]: Consider using a keyword that matches both Event 1 and Event 2, but Event 2 eventually doesn't show up because it's expired. Also consider messing with capitalization to make sure that the search was done case-insensitive.

File: spec/controllers/events_controller_spec.rb [hunk 8]
```diff
@@ -2,16 +2,35 @@
...
+        expect(response).to be_success
+        expect(assigns(:events)).to eq upcoming_events
+        expect(assigns(:keyword)).to be_nil
+      end
     end
 
-    it "is successful and assigns upcoming events" do
-      expect(response).to be_success
-      expect(assigns(:events)).to eq upcoming_events
+    context "keyword is not empty" do
```
Comment (seri) [inline spec/controllers/events_controller_spec.rb:20]: After you refactor to use `Event.search()`, test this with mock objects, similar to the context above.

Comment (seri) [inline spec/controllers/events_controller_spec.rb:20]: Don't forget to test `Event.search()` itself too. And make sure that you consider edge cases such as capitalization, events that match one of keywords but not the other, events that expire etc.

File: spec/factories/venues.rb [hunk 9]
```diff
@@ -2,6 +2,6 @@
   factory :venue do
     name { Faker::Name.name }
     full_address { Faker::Address.street_address }
-    region Region.first
+    region { Region.first }
```
Comment (seri) [inline spec/factories/venues.rb:5]: Okay, to be revisited.

File: spec/support/shared_contexts/events.rb [hunk 10]
```diff
@@ -1,5 +1,5 @@
 shared_context 'three standard events' do
```
Comment (seri) [inline spec/support/shared_contexts/events.rb]: Quote.

File: app/controllers/events_controller.rb [hunk 11]
```diff
@@ -1,6 +1,11 @@
...
+      @events = @events.where(where_clause)
```
Comment (seri) [inline app/controllers/events_controller.rb]: We prefer to be expressive in Ruby. So get rid of `where_clause`. Keep it one line.

File: app/views/events/index.html.slim [hunk 12]
```diff
@@ -3,9 +3,13 @@
...
+    p.no-results
+      | No events found!
+  - else
+    - @events.in_groups_of(2) do |group|
```
Comment (seri) [inline app/views/events/index.html.slim]: Use this: `.in_groups_of(2, false)` to get rid of the `compact` below. Also refactor similar cases.

**Comment by seri** [comment]
@minhchau273 In the cucumber step where you input something to a search box, also assign it to `@keyword =`, and then use that to verify that the keyword shows up again.

File: spec/models/event_spec.rb [hunk 13]
```diff
@@ -88,4 +88,50 @@
       end
     end
   end
+
+  describe ".search" do
```
Comment (seri) [inline spec/models/event_spec.rb:92]: This is some careful testing so that's a good thing, but if we keep up with this standard, it'll become very hard to test something more complicated with more branches. Plus, having too many tests makes the test suite slow.

I suggest that you have only one context, where the keyword and the event names are carefully designed so that all the edges I listed are taken into account.

File: spec/controllers/events_controller_spec.rb [hunk 14]
```diff
@@ -2,16 +2,37 @@
...
+      include_context "three standard events"
```
Comment (seri) [inline spec/controllers/events_controller_spec.rb]: Is this still necessary?

File: spec/controllers/events_controller_spec.rb [hunk 15]
```diff
@@ -2,16 +2,37 @@
...
+
+      let(:keyword) { "keyword" }
```
Comment (seri) [inline spec/controllers/events_controller_spec.rb]: It can be Faker pokemon or whatever.

File: spec/controllers/events_controller_spec.rb [hunk 16]
```diff
@@ -2,16 +2,37 @@
...
+      let(:results) { double }
+
+      before do
+        expect(Event).to receive(:search).and_return results
```
Comment (seri) [inline spec/controllers/events_controller_spec.rb]: You have to test that Event.search was called with the correct argument. This is important.
